ast/ddl.go

```diff
@@ -2510,12 +2510,16 @@ func (n *PartitionMethod) acceptInPlace(v Visitor) bool {
 type PartitionOptions struct {
 	node
 	PartitionMethod
-	Sub         *PartitionMethod
-	Definitions []*PartitionDefinition
+	Sub                  *PartitionMethod
+	Definitions          []*PartitionDefinition
+	IsRemovePartitioning bool
```

Comment:
Don't modify `PartitionOptions`, this structure is used in `CREATE TABLE … PARTITION` as well.

This also makes no sense that placing `REMOVE PARTITIONING` inside `PartitionOptions` to fix the `Restore`, as it is _still_ being wrapped inside an `AlterTableSpec`. You could avoid the comma by changing these lines.

https://github.com/pingcap/parser/blob/e6a8bfdc5d3fe591e32138fd89c2c5b941e6639b/ast/ddl.go#L2069-L2073

Reply:
in mysql8 doc:
`REMOVE PARTITIONING` is a rule of `alter_specification`
see: https://dev.mysql.com/doc/refman/8.0/en/alter-table.html

but in implement of mysql8 yacc:
`REMOVE PARTITIONING` is a kind of `partition_options`
See: https://github.com/mysql/mysql-server/blob/8.0/sql/sql_yacc.yy#L7900

so if we wanna full imitation from the implement of mysql8, it makes sense that modifies `PartitionOptions`.

however, I think it will be better that we follow the mysql8 doc, the syntaxes in mysql8 doc compatible with the implement of mysql8.

Reply:
@leoppro In the YACC implementation `REMOVE PARTITIONING` produces a [`PT_alter_table_remove_partitioning` instance](https://github.com/mysql/mysql-server/blob/4869291f7ee258e136ef03f5a50135fe7329ffb9/sql/parse_tree_nodes.h#L4051-L4057) and the `PARTITION BY` produces a [`PT_alter_table_partition_by` instance](https://github.com/mysql/mysql-server/blob/4869291f7ee258e136ef03f5a50135fe7329ffb9/sql/parse_tree_nodes.h#L4034-L4049), and both classes are subclasses of `PT_alter_table_action` shared by all `AlterTableSpec`-like things like [`PT_alter_table_add_column`](https://github.com/mysql/mysql-server/blob/4869291f7ee258e136ef03f5a50135fe7329ffb9/sql/parse_tree_nodes.h#L3729-L3747). So making `REMOVE PARTITION` an `AlterTableSpec` at AST level *is* implementing like MySQL 8.

The names of the rules inside `parser.y` is irrelevant here.

Reply:
Updated. How about this? PTAL @kennytm @leoppro